tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15716,6 +15749,56 @@ void DiagnoseNodeEntry(Sema &S, FunctionDecl *FD, llvm::StringRef StageName,
                diag::err_hlsl_maxrecordssharedwith_references_invalid_arg);
       }
     }
+
+    // Make sure NodeTrackRWInputSharing attribute cannot be applied to
+    // Input Records that are not RWDispatchNodeInputRecord
+    ArgIdx = 0;
+    while (ArgIdx < FD->getNumParams()) {
+      // is an input record
+      if (hlsl::IsHLSLNodeInputType(ParamTy)) {
+        const ParmVarDecl *ParamDecl = FD->getParamDecl(ArgIdx);
+        hlsl::NodeFlags nodeFlags;
+        if (GetHLSLNodeIORecordType(ParamDecl, nodeFlags)) {
+          hlsl::NodeIOProperties node(nodeFlags);
+          bool hasNodeTrackRWInputSharing = false;
+
+          // determine if the NodeTrackRWInputSharing is an attribute on the
+          // template type
+          if (ParamTy->isStructureOrClassType()) {
+            if (const CXXRecordDecl *CXXRD =
+                    ParamTy.getCanonicalType()->getAsCXXRecordDecl()) {
+
+              if (const ClassTemplateSpecializationDecl *templateDecl =
+                      dyn_cast<ClassTemplateSpecializationDecl>(CXXRD)) {
+
+                auto &TemplateArgs = templateDecl->getTemplateArgs();
+                DXASSERT(
+                    TemplateArgs.size() == 1,
+                    "Input record types need to have one template argument");
+                auto &Rec = TemplateArgs.get(0);
+                clang::QualType RecType = Rec.getAsType();
+                if (RecordDecl *RD = RecType->getAs<RecordType>()->getDecl()) {
```

Comment:
I think this functionality of getting the RecordDecl for the user defined record type of `NodeOutput` or `NodeInputRecord` types should be wrapped in a utility function in `HLSLTypes.cpp`.